File: Neuron.js
const Synapse = require("./Synapse");

const LEARNING_RATE = 1/1 ;

class Neuron {
  constructor(connectedNeurons) {
    this.activation = 0;
    this.synapses = [];
    this.relationToCost = 0;
    this.bias = Math.random() * 10 * (Math.random() < 0.5 ? -1 : 1);
    this.biasGradient = 0;

    for (let i = 0; i < connectedNeurons.length; i++) {
      this.synapses.push(new Synapse(connectedNeurons[i]));
    }
  }

  calculateActivation() {
    this.activation = this.bias;
    for (let synapse in this.synapses) {
      this.activation += this.synapses[synapse].getWeightedValue();
    }
  }

  calculateActivationWithoutBias() {
    this.activation = 0;
    for (let synapse in this.synapses) {
      this.activation += this.synapses[synapse].getWeightedValue();
    }
  }


  getActivationRectified() {
    return Math.max(this.activation, 0);
  }

  getActivationSigmoid() {
    return 1 / (1 + Math.exp(-this.activation));
  }

  applyGradient() {
    this.bias -= this.biasGradient * LEARNING_RATE;
    this.biasGradient = 0;
  }
}

module.exports = Neuron;

File: python/network/Neuron.py
from random import random
from network import Synapse
from math import exp

LEARNING_RATE = 1/1


class Neuron:
    def __init__(self, connected_neurons):
        self.activation = 0
        self.synapses = []
        self.relationToCost = 0
        self.bias = random() * 10 * (-1 if random() < 0.5 else 1)
        self.bias_gradient = 0

        for neuron in connected_neurons:
            self.synapses.append(Synapse(neuron))

    def calculate_activation(self):
        self.activation = self.bias
        for synapse in self.synapses:
            self.activation += synapse.get_weighted_value()

    def calculate_activation_without_bias(self):
        self.activation = 0
        for synapse in self.synapses:
            self.activation = self.activation + \
                synapse.get_weighted_value()

    def get_activation_rectified(self):
        return max(self.activation, 0)

    def get_activation_sigmoid(self):
        return 1 / (1 + exp(-self.activation))

    def apply_gradient(self):
        self.bias = self.bias - self.bias_gradient * LEARNING_RATE
        self.bias_gradient = 0
